chat-android/src/main/java/com/ably/chat/Messages.kt

```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
 )
 
 interface MessagesSubscription : Cancellation {
-    suspend fun getPreviousMessages(queryOptions: QueryOptions): PaginatedResult<Message>
+    suspend fun getPreviousMessages(start: Long? = null, end: Long? = null, limit: Int = 100): PaginatedResult<Message>
 }
 
-class DefaultMessages(
+internal class DefaultMessagesSubscription(
+    private val chatApi: ChatApi,
     private val roomId: String,
-    private val realtimeClient: RealtimeClient,
+    private val cancellation: Cancellation,
+    internal val fromSerialProvider: () -> DeferredValue<String>,
+) : MessagesSubscription {
+    override fun cancel() {
+        cancellation.cancel()
+    }
+
+    override suspend fun getPreviousMessages(start: Long?, end: Long?, limit: Int): PaginatedResult<Message> {
+        val fromSerial = fromSerialProvider().await()
+        val queryOptions = QueryOptions(start = start, end = end, limit = limit, orderBy = NewestFirst)
+        return chatApi.getMessages(
+            roomId = roomId,
+            options = queryOptions,
+            fromSerial = fromSerial,
+        )
+    }
+}
+
+internal class DefaultMessages(
+    private val roomId: String,
+    realtimeChannels: AblyRealtime.Channels,
     private val chatApi: ChatApi,
 ) : Messages {
 
+    private var listeners: Map<Messages.Listener, DeferredValue<String>> = emptyMap()
+
+    private var channelStateListener: ChannelStateListener
+
+    private var lock = Any()
+
     /**
      * the channel name for the chat messages channel.
      */
     private val messagesChannelName = "$roomId::\$chat::\$chatMessages"
 
-    override val channel: Channel
-        get() = realtimeClient.channels.get(messagesChannelName, ChatChannelOptions())
+    override val channel: Channel = realtimeChannels.get(messagesChannelName, ChatChannelOptions())
+
+    init {
+        channelStateListener = ChannelStateListener {
+            if (!it.resumed) updateChannelSerialsAfterDiscontinuity()
+        }
+        channel.on(ChannelState.attached, channelStateListener)
+    }
 
     override fun subscribe(listener: Messages.Listener): MessagesSubscription {
-        TODO("Not yet implemented")
+        val deferredChannelSerial = DeferredValue<String>()
+        addListener(listener, deferredChannelSerial)
+        val messageListener = PubSubMessageListener {
+            val pubSubMessage = it ?: throw AblyException.fromErrorInfo(
+                ErrorInfo("Got empty pubsub channel message", HttpStatusCodes.BadRequest, ErrorCodes.BadRequest),
+            )
+            val data = parsePubSubMessageData(pubSubMessage.data)
```

Comment:
Possible optimisation we can do here - rather than parsing data for every subscription, can have an "internal" subscriber that's put on the channel which then translates once and emits to all "actual" subscribers.

In JS we did this by registering the "internal" subscriber with the realtime channel, then registering user listeners with an `EventEmitter` instance on this class.

Reply:
Looking at code further down, won't this cause duplicate messages on multiple subscriptions?

As multiple instances of `messageListener` will call `emitMessage` which iterates all listeners?

Reply:
Yes, you right, I'll add tests :) I was thinking to use internal eventbus, but decided to do this later, because we need to decide on implementation details. I'll create an issue for this

Reply:
https://github.com/ably-labs/ably-chat-kotlin/issues/18